### ISSUE TITLE
Using computed property for root calendar options causes unnecessary event fetching/rerendering

### ISSUE THREAD

**Issue by arshaw** [fullcalendar/fullcalendar-vue#120]
originally reported by @lllopo on #96:

> ...The events, at least when set to be an ajax source, definitely gets rebound on every options set (I tracked my flickering and overall performance problems to that, btw.). This means no caching and as a consequence - performance problems. This absolutely for sure was not the case in FC4, so it is pretty clear the problem is somewhere in the logic you mentioned - for detecting what is changed and what's not in the options object. I can't say if this is only when the events is an ajax source or in all cases, but it's bad enough, having that ajax source is probably the most popular case anyway. Also I think this probably means, that the same issue probably exists with other options that are a function or a complex JSON object, although I don't have proof for that (gut feeling :-)).

https://codesandbox.io/s/fc5-cumputed-not-reactive-i2gqc?file=/src/DemoApp.vue

> ...It work as follows : the calendar picks a random event form a set of events through a function attached to the events property. So, when initially loaded some event is selected. Then, if you click the "Kick it" link below, this changes a totally unrelated property - namely the headerToolbar. At this point you'll probably notice that the selected event will change, while it shouldn't, of course. This not only shows the problem 2 from my post above, but also proves that it exists if the events is attached to a function and an ajax source (which had to be expected actually).

**Comment by arshaw** [comment]
Just to set the scene for this discussion, here is some pseudo code for how someone might have modified a component's internal state, affecting how FullCalendar renders:

## in v4 (old)

```vue
<script>
  export default {
    data: function() {
      return {
        isCalendarCondensed: false
      }
    },
    computed: {
      calendarHeaderToolbar: function() {
        return !this.isCalendarCondensed
      },
      calendarContentHeight: function() {
        return this.isCalendarCondensed ? 400 : 'auto'
      }
    },
    methods: {
      toggleCondensed: function() {
        this.isCalendarCondensed = !this.isCalendarCondensed
      },
      fetchEvents: function(fetchInfo, callback) {
        someNetworkRequest(fetchInfo.start, fetchInfo.end).then(callback)
      }
    }
  }
</script>
<template>
  <FullCalendar
    :headerToolbar="calendarHeaderToolbar"
    :contentHeight="calendarContentHeight"
    :events="fetchEvents"
  />
  <button @click="toggleCondensed">toggleCondensed</button>
</template>
```

## in v5 (current)

```vue
<script>
  export default {
    data: function() {
      return {
        isCalendarCondensed: false
      }
    },
    computed: {
      calendarOptions: function() {
        return {
          headerToolbar: !this.isCalendarCondensed,
          contentHeight: this.isCalendarCondensed ? 400 : 'auto',
          events: this.fetchEvents
        }
      }
    },
    methods: {
      toggleCondensed: function() {
        this.isCalendarCondensed = !this.isCalendarCondensed
      },
      fetchEvents: function(fetchInfo, callback) {
        someNetworkRequest(fetchInfo.start, fetchInfo.end).then(callback)
      }
    }
  }
</script>
<template>
  <FullCalendar :options="calendarOptions" />
  <button @click="toggleCondensed">toggleCondensed</button>
</template>
```

**Comment by arshaw** [comment]
@lllopo, they way your example departs from my v5 pseudocode is that is specifies `events` as an inline anonymous function. Adapting my example to match what you're doing:

```vue
<script>
  export default {
    data: function() {
      return {
        isCalendarCondensed: false
      }
    },
    computed: {
      calendarOptions: function() {
        return {
          headerToolbar: !this.isCalendarCondensed,
          contentHeight: this.isCalendarCondensed ? 400 : 'auto',
          events: function(fetchInfo, callback) {
            someNetworkRequest(fetchInfo.start, fetchInfo.end).then(callback)
          }
        }
      }
    },
    methods: {
      toggleCondensed: function() {
        this.isCalendarCondensed = !this.isCalendarCondensed
      }
    }
  }
</script>
<template>
  <FullCalendar :options="calendarOptions" />
  <button @click="toggleCondensed">toggleCondensed</button>
</template>
```

**Comment by arshaw** [comment]
If using the above technique, where the `events` property is an inline anonymous function, when `this.isCalendarCondensed` is changed, the function will be re-built.

FullCalendar *tries* to be smart about avoiding rerenders when it knows an option has not changed, but in this case, FullCalendar can't be smart. It's a completely new reference to a new function and FullCalendar can't use any of its cheap comparison logic to know it's the same. Thus, it rerefetches and thus rerenders all events.

To avoid this, it's best store this function as a method of the component, as shown in my first comment at the top of the thread.

The other solution is to make `fetchEvents` a top-level function within the file, like this:

```vue
<script>
  export default {
    ...
    computed: {
      calendarOptions: function() {
        return {
          headerToolbar: !this.isCalendarCondensed,
          contentHeight: this.isCalendarCondensed ? 400 : 'auto',
          events: fetchEvents
        }
      }
    },
    ...
  }

  function fetchEvents(fetchInfo, callback) {
    someNetworkRequest(fetchInfo.start, fetchInfo.end).then(callback)
  }
</script>
```

**Comment by arshaw** [comment]
It's easy to make the mistake of creating an inline anonymous function for your `events` that inadvertently rerenders all events. As @lllopo will no doubt agree, v5 makes it MUCH easier to fall into this trap because it encourages you to make `calendarOptions` one big computed property.

I'll leave this ticket open to discuss potentially reverting back to the old API of having individual props for individual options. This would only happen for v6 if it happens at all.

In the meantime, I should definitely write some docs to prevent other ppl from getting confused by this.

**Comment by lllopo** [comment]
@arshaw Yeah ... I get what you mean and honestly - it also crossed my mind this could be the thing. Yet - the same happens when FC5 is configured with an ajax source of the sorts :
```
events:              {
						url:         "/whatever.php",
						method:      "POST",
						extraParams: function() {
							return {
								"cachebuster": new Date().valueOf()
							};
						},
						failure:     this.ajaxLoadFailed
					}
```

I believe this is a very common code construct and hardly anyone would extract the ajax json object, rather than just put it in place. So, it is really something that definitely needs to be addressed or at least - very, very well documented (big red letters :-)), so no one falls in this trap.

**Comment by kgilden** [comment]
The way I had solved it (and already back in v4, because I was doing `<Fullcalendar v-bind="options" />`) was to simply not make options a computed prop. Having it as a data prop sidesteps this problem.

I'd be :+1: with reverting back to the old style in v6, but only if all event listeners could be passed as props too instead of having to specify them as event listeners. I think that was quite bad DX in v4, because you'd have to more or less use trial and error to figure out if it should be `:option="foo"` or `@option="foo"`.

**Comment by Surt** [comment]
Hi, I'm having a difficult time trying to resolve this on v.5
Actually, I did set the ```calendarOptions``` as ```data``` with events computed in it:

`  data() {
    return {
      calendarOptions: {
        plugins: [
          dayGridPlugin,
          timeGridPlugin,
        ],
        headerToolbar: {left: 'prev,next', center: 'title', right: 'dayGridWeek,dayGridMonth'},
        initialView: 'dayGridMonth',
        events: this.events,
        weekends: true,
        locales: allLocales,
        locale: this.$i18n.locale,
        eventMouseEnter: this.handleEventMouseEnter,
        datesSet  : this.search
      },
      modalData: undefined,
      modalVisible: false,
      editModal: false,
      dateViewStart: '',
      dateViewEnd: '',
    }
  },
`

events computed is a processing of a vuex state. But nothing renders. Triying to avoid the infinite "re-render" due to computed ```calendarOptions``` I make a mess and now I'm completely lost :) 

Any advice? How can I make ```events``` to be a "computed" but without the calendar re-renders on loop due to `datesSet` event.

Thanks in advance!
